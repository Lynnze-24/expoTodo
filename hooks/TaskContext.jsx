import React, { createContext, useContext, useState,useEffect} from 'react'
import uuid from 'react-native-uuid';
import AsyncStorage from '@react-native-async-storage/async-storage';

const TaskContext = createContext();


export function useTaskContext(){
    return useContext(TaskContext)
}

export function TaskProvider({children,mock,val}) {
    // mock and val are for testing
    const initial = mock?mock : [];
    const [task, setTask] = useState(initial);
    const fetchTasks = mock||val? ()=>{return}:async() => {
        let localTasks= await AsyncStorage.getItem('TaskAHL') || null;
        if(localTasks){
            setTask(JSON.parse(localTasks));
        }
    }// second function is for testing 

    
    useEffect(()=> {
        fetchTasks()
    
    }, [])
    
    const addToTask = async(item)=>{
        let newTask = [...task,item]
        setTask(newTask)
        AsyncStorage.setItem('TaskAHL',JSON.stringify(newTask))
    }
    const removeFromTask = async(id)=>{
        let newTask = task.filter(x=> x.id!== id);
        setTask(newTask)
        AsyncStorage.setItem('TaskAHL',JSON.stringify(newTask))
    }

    const toggleCheck = async(id) =>{
        let newTask = [...task];
       let changedItem = newTask.find(x=> x.id=== id);
       changedItem.isChecked = !changedItem.isChecked;
       setTask(newTask);
       AsyncStorage.setItem('TaskAHL',JSON.stringify(newTask))
    }
    
    const value = val || {task,addToTask,removeFromTask,toggleCheck};
    
    return (
        <TaskContext.Provider value={value}>
        {children}
      </TaskContext.Provider>
    )
}