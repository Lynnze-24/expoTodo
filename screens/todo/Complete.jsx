import React from 'react'
import { View, Text,StyleSheet,FlatList} from 'react-native'
import EmptyTask from '../../components/EmptyTask'
import Task from '../../components/Task'
import TaskContainer from '../../containers/TaskContainer'
import { useTaskContext } from '../../hooks/TaskContext'



const Complete = () => {
    const renderItem = ({ item }) => <Task name={item.name} isChecked={item.isChecked} id={item.id} key={item.id}/>;
    const {task} = useTaskContext()
    const completedTasks = task.filter(x=> x.isChecked)
    return (
       <TaskContainer  testID='CompleteScreen'>
           {completedTasks.length?
           (<FlatList data={completedTasks} renderItem={renderItem} keyExtractor={item => item.id} />)
           :(<EmptyTask text="There is no completed task" />)
        }
           
            
        </TaskContainer>
    )
}




export default Complete
