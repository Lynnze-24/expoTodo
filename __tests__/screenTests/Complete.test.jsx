import { render } from '@testing-library/react-native';
import Complete from '../../screens/todo/Complete';
import { TaskProvider } from '../../hooks/TaskContext';

// mock  functions 
const addToTask = jest.fn();
const task = [{
    id:'aaa',
    isChecked:false,
    name:'go shopping'
},{
    id:'bbb',
    isChecked:true,
    name:'read books'
}];
const removeFromTask = jest.fn();
const toggleCheck = jest.fn();

import renderer from 'react-test-renderer';
it('renders correctly', () => {
    const comp = renderer.create((<TaskProvider >
        <Complete/>
     </TaskProvider>)).toJSON();
    expect(comp).toMatchSnapshot();
});


it('should render component correctly',()=>{
    // will render both checked and unchecked
    const { getByText,getByTestId,queryByText}  = render(
           <TaskProvider val={{task,addToTask,removeFromTask,toggleCheck}}>
              <Complete/>
           </TaskProvider>
            );
             
             const completedtask= getByText('read books')
             getByTestId('CompleteScreen')
             expect(completedtask).toHaveStyle({
                textDecorationLine:'line-through'
             })// checked so should have completed style
             expect(queryByText('go shopping')).toBe(null); // should not render incomplete task

})

it('should render empty text component when there is no task',()=>{
    
    const task = [];
    const { getByText,getByTestId}  = render(
           <TaskProvider mock={true} val={{task,addToTask,removeFromTask,toggleCheck}}>
             <Complete/>
           </TaskProvider>
            );
            getByTestId('emptyTask')
            getByText('There is no completed task')

})
