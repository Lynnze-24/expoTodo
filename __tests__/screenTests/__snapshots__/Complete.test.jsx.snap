// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<View
  style={
    Object {
      "backgroundColor": "#fffeee",
      "flex": 1,
      "paddingHorizontal": 20,
    }
  }
  testID="CompleteScreen"
>
  <View
    style={
      Object {
        "alignItems": "center",
        "flex": 1,
        "justifyContent": "center",
      }
    }
    testID="emptyTask"
  >
    <Text>
      There is no completed task
    </Text>
  </View>
</View>
`;
